### DISCUSSION TITLE
No pingpong page migration due to IsPinned

### DISCUSSION THREAD

**Discussion by kingjunsung** [sarchlab/mgpusim#97] [Q&A]
Dear Mr. Sun,
I am currently working on implementing a baseline where the same page migrates back and forth among multiple GPUs. However, I have encountered an issue in the mmu.go code. After the initial migration is performed, the page becomes pinned in a first-touch manner, preventing any further migrations (I guess due to "page.IsPinned" variable). When I comment out the isPinned code and run the application, most of the applications fail to execute properly. Could you please advise on how to enable continuous page migrations without the page being pinned after the first migration?

**Comment by syifan** [comment]
This is not easy due to deadlock prevention requirements. It is likely that one page is migrated to the destination, and before the read can actually happen, is being migrated to another GPU. So, to guarantee forward progress, we pin pages after the first migration. Another way to support continuous page migration is to pin the page for a while before it can be migrated.
